src/Polysemy/Fish.hs

```diff
@@ -0,0 +1,282 @@
+-- literal.
+type family MissingListQuote (type_ :: k) :: a where
+  MissingListQuote t = TypeError (
+         Text "Unexpected list type: " :<>: ShowType t
+    :$$: Text "Perhaps you meant a type-level list? (with quote): '" :<>: ShowType t
+    )
+
+-- | Called by 'ToEffList' when supplied type is not convertible into list of
+-- effects.
+type family ExpectedEffListMonadFound (x :: k) :: a where
+  ExpectedEffListMonadFound x = TypeError (
+      Text "Expected effect, list of effects or monad, found '" :<>: ShowType x :<>: Text "'"
+    )
+
+
+------------------------------------------------------------------------------
+-- Effect list construction --------------------------------------------------
+------------------------------------------------------------------------------
+-- | Converts value into list of effects to be used by other combinators.
+type family ToEffList (caller :: Symbol)
+                      (es :: k)
+                   :: [(Type -> Type) -> Type -> Type] where
+  -- Single effect to wrap in list -------------------------------------------
+  ToEffList _ (e :: (Type -> Type) -> Type -> Type)    = '[e]
+  -- We need concrete types to get GHC to reduce our type family
+  ToEffList c (e :: k -> l -> Type)                    = MissingEffTypes c e
+  -- List of effects ---------------------------------------------------------
+  -- Empty case to help type inference
+  ToEffList _ '[]                                      = '[]
+  ToEffList _ (es :: [(Type -> Type) -> Type -> Type]) = es
+  ToEffList c (es :: [k -> l -> Type])                 = MissingEffTypes c es
+  -- We do not want to pretend list type is valid value --- instead we throw
+  -- an error
+  ToEffList _ []                                       = MissingListQuote []
+  ToEffList _ [t]                                      = MissingListQuote [t]
+  -- Single 'Monad' to wrap in 'Lift' and list -------------------------------
+  ToEffList _ (m :: Type -> Type)                      = '[Lift m]
+  -- Unexcpected value -------------------------------------------------------
+  ToEffList _ (x :: k)                                 = ExpectedEffListMonadFound x
+
+-- | Extracts list of effects from 'Sem'.
+type family SemList s where
+  SemList (Sem r _) = r
+
+
+------------------------------------------------------------------------------
+-- "Fish" operators ----------------------------------------------------------
+------------------------------------------------------------------------------
+-- | Infix equivalent of 'Sem' with automatic conversion of supplied effects
+-- into list. Use ('>@>') instead if you only want to work with 'Members' of
+-- your union.
+--
+-- __Examples:__
+--
+-- 'Sem' with list of multiple effects:
+--
+-- @
+-- foo :: 'Sem' \'['Polysemy.State.State' 'Int', 'Polysemy.Input.Input' 'String'] ()
+-- @
+--
+-- can be written as:
+--
+-- @
+-- foo :: \'['Polysemy.State.State' 'Int', 'Polysemy.Input.Input' 'String'] \@> ()
```

Comment:
I'm not sure this is a good example, ideally people will never write a `Sem` like that.

Reply:
Maybe show `State Int ': r` instead?

Reply:
Ok, will do.